### PR TITLE
ProgPoW Review

### PR DESCRIPTION
**Pull request by gcolvin**
This is a PR for review ahead of All Core Devs call 96.

We Do Not recommend that this Proposal be deployed at this time. Rather it is being offered in the spirit of [Ben DiFrancesco's compromise](https://ethereum-magicians.org/t/a-progpow-compromise-pre-proposal/4057) which for our purposes we state simply as,
* This Proposal is not being proposed for deployment in any planned hardfork.
* This Proposal should be fully implemented and tested across major clients.
* Clients implementing this Proposal should be deployed and maintained on a testnet.
* This leaves open the possibility and threat of future deployment.

Incorporates @AndreaLanfranchi's [implementation](https://github.com/ifdefelse/ProgPOW/pull/52) of the [Kik exploit](https://github.com/kik/progpow-exploit) fix and a start at @bitsbetrippin client, exchange, and pool tracking tables.

The reference implementation is on @ifdefelse's repo at https://github.com/ifdefelse/ProgPOW

### CODE REVIEW

**Comment by gcolvin** [comment]
Thanks for your thorough review, @MicahZoltu.

This EIP has been through a lot since it was first proposed in 2018, and non-technical issues have loomed larger than technical ones.  So I'd rather leave the Simple Summary and Motivation sections be.  The Simple Summary is what we are proposing upfront that the Core Devs do now, and needs to somewhere.  And the Motivation is why we propose to do it now -- if we don't it will soon enough it be irrelevant.

The Specification should mostly be pruned as you suggest - the historical information can be found in the @ifdefelse repo. The material on the audits, the Kik exploits, and such should remain in the document - probably as Security Considerations - as they are important to those who have been following the proposal and want to judge whether changes based on the audits and Kik exploit are sound.  I'm not sure whether copyright issues allow us to move the audit documents to ../assets.

I've reduced the Example/Test Cases section down to just the Test Cases for this latest version.

The tables in the Implementation section are meant for Michael Carter (@bitsbetrippin) to track the status of the EIP.  I'll leave it to him to better explain their meaning -- but as I understand it they show that months ago most clients had implemented and tested the software and the pools and exchanges were ready to fork to it.  A political statement, perhaps.  I'll leave it to Michael and James Hancock (MadeofTin) to decide if they belong here or elsewhere.  And the opening lines of the Implementation section are the same ones that you want removed from the Simple Summary.  For most any other EIP I'd agree, but this is not an ordinary EIP.  I have moved both these lines and the table to a non-canonical Deployment section for now.

**Comment by gcolvin** [comment]
@MicahZoltu I have changed all but the first sentence of the Simple Summary to a _Note_.

**Comment by MicahZoltu** [comment]
I'll wait for other editors to comment on this, and will ultimately defer to them on merging.

I'm still pretty apprehensive about the statements regarding deployment/recommendations in the EIP (rather than discussion-to link).  I don't believe that any EIP should ignore our traditional stance that "EIPs are technical specifications, free from politics and opinions".  No Draft/Accepted EIP is guaranteed to show up on mainnet, and we don't say that at the top of every Draft/Accepted EIP so I don't think it should be included here just because some trolls on the internet don't understand the EIP process.

I also still dislike the references to "versions" in the latter sections.  I would rather see this worded something like:
> An early draft of this specification did [X], but it was found to be vulnerable to [Y] so we changed the specification to [Z].

The above feels more appropriate for the Rationale section, though I can appreciate the reasoning for putting it in the Security Considerations section.  For security considerations section I would rather see discussion of known unknowns or something, rather than vulnerabilities in early drafts that were fixed before this EIP was finalized.

----

Personal Note: Despite my nagging about "EIPs are technical specifications", I think the EIP is well written and appropriately detailed overall.

**Comment by gcolvin** [comment]
@MicahZoltu I share your apprehensions.

I can replace the reference to version 0.9.3 with a link to the corresponding PR on the reference implementation, but the reader does need to know where to find the change.  And I do think the change is a security consideration.

Personal note: I actually had not much to do with most of the writing here, so I'll accept you praise on the other authors behalf.

**Comment by MicahZoltu** [comment]
> the reader does need to know where to find the change

Why does a reader implementing this in the future need to know where to find a change to the specification prior to it being final?

**Comment by gcolvin** [comment]
@MicahZoltu 
> Why does a reader implementing this in the future need to know where to find a change ...

In the future, they may not.  Currently, some consider the existence of the Kik exploit and/or the changes made to fix it as a reason to reject the proposal.  So reviewers need to easily find the discussions of the exploit, its (lack of) severity, how it was fixed, etc. to judge whether they agree.

File: EIPS/eip-1057.md [hunk 1]
```diff
@@ -455,7 +433,7 @@ The flow of the overall algorithm is:
 * When mining this final value is compared against a `hash32_t` target
 
 ```cpp
-hash32_t progPowHash(
+ash32_t progPowHash(
```
Comment (chfast) [inline EIPS/eip-1057.md:436]: Definitely a backdoor sneaked in.

Comment (gcolvin) [reply EIPS/eip-1057.md:436]: good catch

**Comment by MicahZoltu** [comment]
> Currently, some consider the existence of the Kik exploit and/or the changes made to fix it as a reason to reject the proposal. So reviewers need to easily find the discussions of the exploit, its (lack of) severity, how it was fixed, etc. to judge whether they agree.

I feel like the solution to the problem of people not reading specifications before rejecting them is not one we should solve by bloating the specification with history.  All history is in git, and if you want to see discussion about changes you can go to the discussion-to link.

I'm pretty strongly in the camp that EIPs should be pure technical specifications and should be written for future (5-10 years) readers, not people who are knee deep in discussions and debates occurring around the time the specification was written.

As I mentioned before, I'll defer to the judgement of other editors, but I personally would want to see something more compelling than "people aren't actually reviewing the spec before rejecting it" to be OK with including all of this historic stuff in the EIP itself.

**Comment by Souptacular** [comment]
I'm torn on whether or not political/historical information needs to be included in the EIP. I want to defer to the other editors, but I'm not sure how many other editors are going to see this (maybe @axic?). I'm leaning towards keeping in the parts about the history and deployment information since it's a net positive to have that information centralized so people don't have to scroll through Git commits and EthMag's forum posts to find the latest.

**Comment by lightclient** [comment]
Not an editor, but I also feel strongly that Standards Track EIPs should not include political/historical information. I would, however, welcome an informational EIP with this information with open arms.

**Comment by gcolvin** [comment]
I hear and share all apprehensions.  In this case, these notes on deployment and history can be removed to another place before deployment, but I think the EIP needs them for now so that any reader can understand the EIP in context.

**Review by que89 — APPROVED** [review]

**Comment by lightclient** [comment]
@gcolvin this is still a draft PR, are you still working on this or are you waiting for it to be merged? It looks like this should be addressed at minimum: https://github.com/ethereum/EIPs/pull/2894#discussion_r478729907.

**Comment by gcolvin** [comment]
@lightclient  Yes, it's a Draft PR -- cannot be merged, still being worked on.  The typo that @chfast caught is fixed downstream.

**Comment by gcolvin** [comment]
@MicahZoltu @Souptacular 

@AndreaLanfranchi is healthy and looking at the Kik exploit again, and it's becoming the consensus of the authors that it and similar exploits are not at all practical to execute in the real world.  Andrea can explain better.  That leaves our proposal much simpler, with no technical changes since the audits.  The Kik fix and light-evaluation mitigation will still need to be in the security section, but not in so much detail.  Ethash is also vulnerable to the light-evaluation attack, so mitigating it a topic for separate discussion.  For ProgPoW changing one constant is all that is required.

**Comment by MicahZoltu** [comment]
@gcolvin I believe we are all waiting for this to move out of draft before dedicating time to really giving it a deep dive.  Also, I doubt my stance is going to change much, but others may have different views on the matter.

The motivation section also feels too long.  As hinted previously, I think most of that should be moved to a discussion link.  Ideally, the motivation section would give a high level overview of why someone may want to use this EIP but it doesn't need to go into extensive detail about all of the arguments for/against the EIP, all of the possible alternatives, etc.  The EIP should really just be a technical standard and the purpose of the motivation section is to give a little bit of context to future readers, not a history lesson.

**Comment by MicahZoltu** [comment]
It is worth noting that there is now the https://github.com/ethereum/eth1.0-specs/ repository where it is more appropriate to discuss "what is going to go into a hard fork".  A large amount of what is in this EIP I think may be more appropriate there.

**Comment by AndreaLanfranchi** [comment]
@gcolvin just wrote a [detailed explanation](https://github.com/ifdefelse/ProgPOW/issues/51#issuecomment-690155355) about why Kik's finding is **impossible** to exploit and all the noise around has no foundation.

**Comment by MicahZoltu** [comment]
To be clear to readers who may not be familiar with the EIP process: I **want** to merge this EIP as a draft.

However, I am unwilling to go against my standard EIP merging process for this PR which is that EIPs are purely technical standards, not a place to record history, record discussions, have arguments, sell people on the merits of a thing, convince people to use a thing, etc.  If this PR was just a technical specification of ProgPoW with a few paragraphs of motivation to give context and a rationale for why certain choices were made over alternatives then I would not push back against merging this at all.

It is worth noting that the other aspect of my standard EIP merging process is to defer to the larger EIP editor group even when I disagree, so if enough other active editors think we should merge this with all of the non-technical-stuff in it, then I'll go along with the group.

**Comment by Eliovp** [comment]
Why is this all still ongoing? 

The initial idea/proposal here, "ProgPow", was basically to keep out ASICS or at least to slow them down? 
Or am I wrong?

Is there a problem with how things currently are?
Are the GPU miners so badly affected? Hell no! On the contrary, things are getting better on a daily base.

The only thing that will change by forking to ProgPow is this:
"**Hello dear FPGA's, we welcome you because we now made it much easier for you to dominate the network! Welcome!"**
Oh, Right, and Nvidia based GPU's taking the upper hand. Much politics?

In all seriousness, why go through all this trouble and why in God's name would you even think about risking the fact that it might go completely sideways... ??

Might be me though 🤷‍♂️

**Comment by gcolvin** [comment]
@MicahZoltu @Souptacular  We authors have come to consensus that the Kik exploit is not just impractical to deploy, but impossible.  And we can devise a similar vulnerability that avoids Kik's mistake, but it is uses an even more impractical amount of resources.  So we backed out the most recent changes, _leaving a single specification_.

I've moved a brief description of the vulnerabilities with links to full discussions to the Security Considerations section.  We also describe fixes, and argue that they should not be deployed.

**Comment by gcolvin** [comment]
@MicahZoltu 
>However, I am unwilling to go against my standard EIP merging process for this PR which is that EIPs are purely technical standards, not a place ...

What would we need to change to meet your standard?

**Comment by MicahZoltu** [comment]
> What would we need to change to meet your standard?

Things I would want to see changed to feel comfortable merging this and helping push this EIP through to final:

Remove everything from the Simple Summary except:
> A new Proof-of-Work algorithm to replace Ethash that utilizes almost all parts of commodity GPUs.

Remove the versioning stuff in the specification (the table) and just state what this specification specifies.  The values could be kept in a separate list as they are now but without the header row and with only one column, or they could just be inlined into the definitions above the table to reduce repetition in the EIP.  Also remove the lead-in line for those tables:
> The values of these parameters have been tweaked between the original version and the version proposed here for Ethereum adoption. See this medium post for details.

From Security Considerations remove:
> We do not recommend implementing this fix at this time.

From Security Considerations remove:
> This fix is available as a PR to the reference implementation. Again, we do not recommend implementing this fix at this time. Kik's vulnerability and others like it cannot be exploited now and likely never will be. It's better to deploy the audited code.

Remove the entire Implementation section *except* the first line.

Change the `License and Copyright` section to just `Copyright` (following EIP template) and remove the mention of the license of an externally linked product (that product should include a license file).  We don't specify licenses for externally linked content anywhere else in the EIPs repository, and I don't think we should change that policy here or introduce such a policy.

----

It is worth noting that there is an upcoming change to the EIP process that will fully separate it from the Ethereum hardfork inclusion process, which means that this EIP could go all the way to final without ever needing to answer the question of "will this be included in Ethereum mainnet".

**Comment by gcolvin** [comment]
@MicahZoltu 
> From ... remove ...

Right.  The notes on deployment will either become a recorded ACD Decision soon or else the proposal will be withdrawn -- either way they will be removed from the Draft.

I do think what remains about the Kik exploit is about right.  Will look at trimming.

I'll leave the implementation notes to @BitsBeTrippin and @MadeofTin.

> It is worth noting that there is an upcoming change to the EIP process that ...

Yes.  This proposal has endured more than one such...

**Comment by gcolvin** [comment]
@MicahZoltu I have decided exercise to self-merge this, as Martin is waiting on result, and I think we have resolved as much as we can.  I've removed everything but a part of the Kik discussion that I think needs to remain.  It was Accepted long ago, so this isn't like this is the first draft to be merge.  Thanks again for your assistance!

**Pull request merged**